bpo-36887: add math.isqrt

mdickinson opened the pull request:
Add an integer square root function to the math module.

<!-- issue-number: [bpo-36887](https://bugs.python.org/issue36887) -->
https://bugs.python.org/issue36887
<!-- /issue-number -->

mdickinson committed:
Add math.isqrt function computing the integer square root.

serhiy-storchaka commented on Modules/mathmodule.c:
```diff
@@ -1476,6 +1476,243 @@ count_set_bits(unsigned long n)
+/*[clinic input]
+math.isqrt
+
+    n : 'O'
```
```suggestion
    n: object
```

mdickinson commented on Modules/mathmodule.c:
```diff
@@ -1476,6 +1476,243 @@ count_set_bits(unsigned long n)
+    size_t n_length, c, e, d, temp;
+    PyObject *a, *q, *n_shifted, *a_temp, *shift;
+
+    /* Hmm. Problem. Doesn't convert True to 1. */
```
These two lines should be deleted.

mdickinson commented on Modules/mathmodule.c:
```diff
@@ -1476,6 +1476,243 @@ count_set_bits(unsigned long n)
+        return NULL;
+    }
+
+    /* references owned: n */
```
We can probably get rid of these comments, too; I found them useful while figuring out what I needed to `DECREF` in the various failure branches, but they've served their purpose now.

serhiy-storchaka commented on Modules/mathmodule.c:
```diff
@@ -1476,6 +1476,243 @@ count_set_bits(unsigned long n)
+
+    if (n_sign == 0) {
+        Py_DECREF(n);
+        a = _PyLong_Zero;
```
```c
return PyLong_FromLong(0);
```
is shorter. `_PyLong_Zero` is convenient if you do not need a new reference and want to omit the error check.

mdickinson replied:
Good point; I'll use that.

mdickinson commented on Modules/mathmodule.c:
```diff
@@ -1476,6 +1476,243 @@ count_set_bits(unsigned long n)
+    n : 'O'
+    /
+
+Return the largest integer not exceeding the square root of the input.
```
This is a bit convoluted. "Return the integer part of the square root of the input." may be more direct and accessible.

mdickinson committed:
Code cleanup: remove redundant comments, rename some variables.

mdickinson committed:
Tighten up code a bit more; use Py_XDECREF to simplify error handling.

mdickinson commented on Modules/mathmodule.c:
```diff
@@ -1476,6 +1476,194 @@ count_set_bits(unsigned long n)
+
+    https://github.com/mdickinson/snippets/blob/master/proofs/isqrt/src/isqrt.lean
+
+The comments in that file also include an informal proof. Note that the
```
I'll add a sketch of the proof in the comments here, to eliminate the reliance on an external source.

mdickinson replied:
Done.

serhiy-storchaka commented on Modules/mathmodule.c:
```diff
@@ -1476,6 +1476,194 @@ count_set_bits(unsigned long n)
+{
+    int a_too_large, s;
+    size_t n_bit_length, c, d, e;
+    PyObject *a=NULL, *b, *q=NULL, *shift;
```
```suggestion
    PyObject *a = NULL, *b, *q = NULL, *shift;
```

mdickinson commented on Doc/library/math.rst:
```diff
@@ -166,6 +166,14 @@ Number-theoretic and representation functions
+.. function:: isqrt(n)
+
+   Return the "integer square root" of the integer *n*. This is the largest
+   integer *s* not exceeding the exact square root of *n*.
```
From comments on the issue: add the recipe `1 + isqrt(n - 1)` for computing the *ceiling* of the square root of a positive integer.

mdickinson replied:
Done.

serhiy-storchaka commented on Modules/mathmodule.c:
```diff
@@ -1476,6 +1476,194 @@ count_set_bits(unsigned long n)
+        Py_DECREF(a);
+        Py_DECREF(shift);
+        if (b == NULL) {
+            goto error;
```
Double decref of `a`.

mdickinson replied:
The second DECREF is an XDECREF, so I think this is okay.

mdickinson replied:
> so I think this is okay.

No, it's not. Now fixed.

serhiy-storchaka commented on Modules/mathmodule.c:
```diff
@@ -1476,6 +1476,194 @@ count_set_bits(unsigned long n)
+        Py_DECREF(b);
+        Py_DECREF(q);
+        if (a == NULL) {
+            goto error;
```
Double decref of `q`.

mdickinson replied:
Same here: the error branch uses `Py_XDECREF` for `q`.

mdickinson replied:
But I realise I missed your point: I could move the `Py_DECREF(q)` to after the error check. I'll do that.

mdickinson replied:
On second thoughts; I'll leave as is - I find it simpler to read and write this way (consistently DECREF'ing arguments before doing the error check).

serhiy-storchaka replied:
`Py_XDECREF` does not help because `q` is not NULL. You should either set `q = NULL` after `Py_DECREF(q)` or use `Py_CLEAR(q)` instead of `Py_DECREF(q)` or remove `Py_XDECREF(q)` after the `error` label and add `Py_DECREF(q)` before corresponding `goto error`.

The same is for `a`.

mdickinson replied:
Yep, I'm an idiot. Will fix shortly. Thank you!

mdickinson replied:
Should be fixed now.

serhiy-storchaka commented on Modules/mathmodule.c:
```diff
@@ -1476,6 +1476,194 @@ count_set_bits(unsigned long n)
+    }
+
+    if (a_too_large) {
+        b = PyNumber_Subtract(a, _PyLong_One);
```
In one line:
`Py_SETREF(a, PyNumber_Subtract(a, _PyLong_One));`

mdickinson replied:
Ah, that's definitely nicer. Will change.

mdickinson committed:
Update Modules/mathmodule.c

Co-Authored-By: Serhiy Storchaka <storchaka@gmail.com>

mdickinson committed:
Update Modules/mathmodule.c

Use real argument clinic type instead of an alias

Co-Authored-By: Serhiy Storchaka <storchaka@gmail.com>

mdickinson committed:
Add proof sketch

mdickinson committed:
Merge remote-tracking branch 'mdickinson/math-isqrt' into math-isqrt

mdickinson committed:
Updates from review.

mdickinson commented on Doc/library/math.rst:
```diff
@@ -166,6 +166,18 @@ Number-theoretic and representation functions
+
+   For some applications, it's convenient to have the ceiling of the
+   square root of *n* rather than the floor. For a positive *n*, the
+   ceiling of the square of *n* can be computed with `1 + isqrt(n - 1)`.
```
*square root*, not *square*

mdickinson replied:
Also, need double backticks around the code snippet.

mdickinson committed:
Correct and expand documentation.

mdickinson committed:
Fix bad reference handling on error; make some variables block-local; other tidying.

mdickinson committed:
Style and consistency fixes.

serhiy-storchaka commented on Modules/mathmodule.c:
```diff
@@ -1476,6 +1476,267 @@ count_set_bits(unsigned long n)
+        ++s;
+    }
+
+    a = PyLong_FromLong(1);
```
Check for error.

mdickinson replied:
Done. Thanks.

serhiy-storchaka commented on Modules/mathmodule.c:
```diff
@@ -1476,6 +1476,267 @@ count_set_bits(unsigned long n)
+    return a;
+
+  error:
+    Py_DECREF(a);
```
`a` can be NULL.

mdickinson replied:
Sigh. Now I remember why I like coding in Python ...

Thanks; fixed.

mdickinson committed:
Add missing error check; don't try to DECREF a NULL a

mdickinson committed:
Simplify some error returns.

mdickinson committed:
Another two test cases:

- clarify that floats are rejected even if they happen to be
  squares of small integers
- TypeError beats ValueError for a negative float

serhiy-storchaka commented on Doc/library/math.rst:
```diff
@@ -166,6 +166,20 @@ Number-theoretic and representation functions
+   part of the exact square root of *n*.
+
+   For some applications, it is convenient to have the ceiling of the square
+   root of *n* (the smallest integer ``a`` such that ``n <= a * a``) rather
```
If you prefer more typographic display of the formula:
```suggestion
   root of *n* (the smallest integer *a* such that *n* ≤ *a*\ ²) rather
```
If you prefer this, it would be worth to use also non-breakable spaces around the ≤ operator: https://stackoverflow.com/questions/11830242/how-can-i-make-a-non-breaking-space-in-restructuredtext

serhiy-storchaka commented on Doc/library/math.rst:
```diff
@@ -166,6 +166,20 @@ Number-theoretic and representation functions
+
+   For some applications, it is convenient to have the ceiling of the square
+   root of *n* (the smallest integer ``a`` such that ``n <= a * a``) rather
+   than the floor (the largest integer ``a`` such that ``a * a <= n``). For a
```
I afraid that it may be confusing that the actual definition of the function (the largest integer `a` such that `a * a <= n`) goes after the definition of the alternate function. I suggest to move it to the previous paragraph and keep this paragraph for ceiling only.

mdickinson replied:
Agreed. I'll move this.

serhiy-storchaka reviewed (approved):
LGTM. I did not analyze the original formulas, but the C code looks correct.

I left just few minor suggestions to the documentation. You can ignore them.

mdickinson committed:
Documentation and markup improvements; thanks Serhiy for the suggestions!

mdickinson committed:
Cleaner Misc/NEWS entry wording.

mdickinson committed:
Clean up (with one fix) to the algorithm explanation and proof.
